### PR TITLE
75 Introduce Codes to Rules

### LINKED ISSUE #75: Filtering applicable rules from available rules via configuration.

### PR DESCRIPTION
**Pull request by netme**
Fixes #75 

## How to test
1. Add `zally.ignoreRules = M015` to your `application.properties`
2. Run server
3. Run client:
    ```bash
    ./cli/bin/zally ../server/src/test/resources/fixtures/api_spa.yaml 
    ```
4. Observe that "Use Problem JSON" rule is not showing

### CODE REVIEW

File: server/src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ class AvoidJavascriptKeywordsRule : AbstractRule() {
     override val url = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
             "#should-reserved-javascript-keywords-should-be-avoided"
     override val violationType = ViolationType.SHOULD
+    override val code = "S001"
```
Comment (mkulak) [inline server/src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.kt]: Remove TODO block in lines 7-11, add `@Component` annotation, disable it via `application.properties` and delete `DisabledAvoidJavascriptKeywordsRuleTest`. Whole purpose of this TODO was to disable this rule. Now we have better mechanism for this.

File: server/src/main/java/de/zalando/zally/rules/RulesValidator.kt [hunk 2]
```diff
@@ -20,6 +20,10 @@ class RulesValidator(@Autowired val rules: List<Rule>) {
             return listOf(Violation(InvalidSwaggerFileDummyRule(), "Can't parse swagger file",
                     e.toString(), ViolationType.MUST, "", emptyList()))
         }
-        return rules.map { it.validate(swagger) }.filterNotNull().sortedBy { it.violationType }
+        return rules
```
Comment (fmueller) [inline server/src/main/java/de/zalando/zally/rules/RulesValidator.kt]: Kotlin is beautiful. Nice to read. :-)

**Comment by netme** [comment]
@fmueller really good idea, I will do it.

File: server/src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.kt (hunk 1 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.kt]: @mkulak , fixed in 201463d

**Comment by netme** [comment]
@fmueller , resolved in 0da2d29

File: server/src/test/java/de/zalando/zally/RuleUniquenessTest.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    lateinit var rules: List<Rule>
+
+    @Test
+    fun rulesShouldBeUnique() {
+        val duplicatedCodes = rules
+                .groupBy { it.code }
+                .mapValues { it.value.sumBy { 1 } }
+                .filterValues { it > 1 }
+                .map { it.key }
+        assertEquals("Duplicated rules found: " + duplicatedCodes, 0, duplicatedCodes.count())
```
Comment (mkulak) [inline server/src/test/java/de/zalando/zally/RuleUniquenessTest.kt]: I propose to express it more concise:
```
rules
    .groupBy { it.title }
    .filterValues { it.size > 1 }
    .keys
```

Comment (netme) [inline server/src/test/java/de/zalando/zally/RuleUniquenessTest.kt]: Thanks @mkulak , that's something I was looking for.

**Comment by netme** [comment]
@mkulak I've addressed your comment in 	9a8c946

**Comment by fmueller** [comment]
👍

**Pull request merged**